File: C#/Silverlight/source/LSilverlight-0.3.3/Loon.Core.Graphics/LGradation.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Loon.Core.Graphics.OpenGL;
using Microsoft.Xna.Framework;
using Loon.Utils;
using System.Runtime.CompilerServices;

namespace Loon.Core.Graphics
{
    public class LGradation : LRelease
    {

        private static Dictionary<String, LGradation> gradations;

        private Color start;

        private Color end;

        private int width, height, alpha;

        private LTexture drawTexWidth, drawTexHeight;

        private LPixmap drawImgWidth, drawImgHeight;

        public static LGradation GetInstance(LColor s, LColor e, int w, int h)
        {
            return GetInstance(s.Color, e.Color, w, h, 125);
        }

        public static LGradation GetInstance(LColor s, LColor e, int w, int h, int alpha)
        {
            return GetInstance(s.Color, e.Color, w, h, alpha);
        }

        public static LGradation GetInstance(Color s, Color e, int w, int h)
        {
            return GetInstance(s, e, w, h, 125);
        }

        public static LGradation GetInstance(Color s, Color e, int w, int h,
                int alpha)
        {
            if (gradations == null)
            {
                gradations = new Dictionary<String, LGradation>(10);
            }
            int hashCode = 1;
            hashCode = LSystem.Unite(hashCode, s.PackedValue);
            hashCode = LSystem.Unite(hashCode, e.PackedValue);
            hashCode = LSystem.Unite(hashCode, w);
            hashCode = LSystem.Unite(hashCode, h);
            hashCode = LSystem.Unite(hashCode, alpha);
            String key = "" + (hashCode);
            LGradation o = (LGradation)CollectionUtils.Get(gradations, key);
            if (o == null)
            {
                CollectionUtils.Put(gradations, key, o = new LGradation(s, e, w, h, alpha));
            }
            return o;
        }

        private LGradation()
        {

        }

        private LGradation(Color s, Color e, int w, int h, int alpha)
        {
            this.start = s;
            this.end = e;
            this.width = w;
            this.height = h;
            this.alpha = alpha;
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public void DrawWidth(GLEx g, int x, int y)
        {
            try
            {
                if (drawTexWidth == null)
                {
                    LPixmap img = new LPixmap(width, height, true);
                    for (int i = 0; i < width; i++)
                    {
                        img.SetColor(
                                (byte)((start.R * (width - i)) / width
                                        + (end.R * i) / width),
                                (byte)((start.G * (width - i)) / width
                                        + (end.G * i) / width),
                                (byte)((start.B * (width - i)) / width
                                        + (end.B * i) / width), (byte)alpha);
                        img.DrawLine(i, 0, i, height);
                    }
                    drawTexWidth = img.Texture;
                }
                g.DrawTexture(drawTexWidth, x, y);
            }
            catch
            {
                for (int i = 0; i < width; i++)
                {
                    g.SetColor(
                                (byte)((start.R * (width - i)) / width
                                        + (end.R * i) / width),
                                (byte)((start.G * (width - i)) / width
                                        + (end.G * i) / width),
                                (byte)((start.B * (width - i)) / width
                                        + (end.B * i) / width), (byte)alpha);
                    g.DrawLine(i + x, y, i + x, y + height);
                }
            }
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public void DrawHeight(GLEx g, int x, int y)
        {
            try
            {
                if (drawTexHeight == null)
                {
                    LPixmap img = new LPixmap(width, height, true);
                    for (int i = 0; i < height; i++)
                    {
                        img.SetColor((byte)(
                                (start.R * (height - i)) / height
                                        + (end.R * i) / height),
                                (byte)((start.G * (height - i)) / height
                                        + (end.G * i) / height),
                                (byte)((start.B * (height - i)) / height
                                        + (end.B * i) / height), (byte)(alpha));
                        img.DrawLine(0, i, width, i);
                    }
                    drawTexHeight = img.Texture;
                }
                g.DrawTexture(drawTexHeight, x, y);
            }
            catch
            {
                for (int i = 0; i < height; i++)
                {
                    g.SetColor(
                            (byte)((start.R * (height - i)) / height
                                    + (end.R * i) / height),
                            (byte)((start.G * (height - i)) / height
                                    + (end.G * i) / height),
                            (byte)((start.B * (height - i)) / height
                                    + (end.B * i) / height), (byte)alpha);
                    g.DrawLine(x, i + y, x + width, i + y);
                }
            }
        }


        [MethodImpl(MethodImplOptions.Synchronized)]
        public void DrawWidth(LPixmap g, int x, int y)
        {
            try
            {
                if (drawImgWidth == null)
                {
                    drawImgWidth = new LPixmap(width, height, true);
                    for (int i = 0; i < width; i++)
                    {
                        drawImgWidth.SetColor(
                                (byte)((start.R * (width - i)) / width
                                        + (end.R * i) / width),
                                (byte)((start.G * (width - i)) / width
                                        + (end.G * i) / width),
                                (byte)((start.B * (width - i)) / width
                                        + (end.B * i) / width), (byte)alpha);
                        drawImgWidth.DrawLine(i, 0, i, height);
                    }
                }
                g.DrawPixmap(drawImgWidth, x, y);
            }
            catch
            {
                for (int i = 0; i < width; i++)
                {
                    g.SetColor(
                            (byte)((start.R * (width - i)) / width
                                    + (end.R * i) / width),
                            (byte)((start.G * (width - i)) / width
                                    + (end.G * i) / width),
                            (byte)((start.B * (width - i)) / width
                                    + (end.B * i) / width), (byte)alpha);
                    g.DrawLine(i + x, y, i + x, y + height);
                }
            }
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public void DrawHeight(LPixmap g, int x, int y)
        {
            try
            {
                if (drawImgHeight == null)
                {
                    drawImgHeight = new LPixmap(width, height, true);
                    for (int i = 0; i < height; i++)
                    {
                        drawImgHeight.SetColor(
                                (byte)((start.R * (height - i)) / height
                                        + (end.R * i) / height),
                                (byte)((start.G * (height - i)) / height
                                        + (end.G * i) / height),
                                (byte)((start.B * (height - i)) / height
                                        + (end.B * i) / height), (byte)alpha);
                        drawImgHeight.DrawLine(0, i, width, i);
                    }
                }
                g.DrawPixmap(drawImgHeight, x, y);
            }
            catch
            {
                for (int i = 0; i < height; i++)
                {
                    g.SetColor((byte)(
                            (start.R * (height - i)) / height
                                    + (end.R * i) / height),
                           (byte)((start.G * (height - i)) / height
                                    + (end.G * i) / height),
                            (byte)((start.B * (height - i)) / height
                                    + (end.B * i) / height), (byte)alpha);
                    g.DrawLine(x, i + y, x + width, i + y);
                }
            }
        }

        public static void Close()
        {
            if (gradations == null)
            {
                return;
            }
            foreach (LGradation g in gradations.Values)
            {
                if (g != null)
                {
                    g.Dispose();
                }
            }
            gradations.Clear();
        }

        public void Dispose()
        {
            if (drawTexWidth != null)
            {
                drawTexWidth.Destroy();
            }
            if (drawTexHeight != null)
            {
                drawTexHeight.Destroy();
            }
            if (drawImgWidth != null)
            {
                drawImgWidth.Dispose();
            }
            if (drawImgWidth != null)
            {
                drawImgWidth.Dispose();
            }
        }
    }
}

File: Java/Loon-Lite(PureJava)/Loon-Lite-Core/src/loon/canvas/LGradation.java
/**
 * Copyright 2008 - 2011
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @project loon
 * @author cping
 * @email：javachenpeng@yahoo.com
 * @version 0.1
 */
package loon.canvas;

import loon.LRelease;
import loon.LSystem;
import loon.LTexture;
import loon.opengl.GLEx;
import loon.utils.IntMap;

/**
 * Color渲染Image的简单渐变实现
 */
public class LGradation implements LRelease {

	private static IntMap<LGradation> COLOR_GRADATIONS;

	private LColor startColor;

	private LColor endColor;

	private int width, height, alpha;

	private LTexture drawTexWidth, drawTexHeight;

	private Image drawImgWidth, drawImgHeight;

	public static LGradation getInstance(LColor s, LColor e, int w, int h) {
		return getInstance(s, e, w, h, 125);
	}

	public static LGradation getInstance(LColor s, LColor e, int w, int h, int alpha) {
		if (COLOR_GRADATIONS == null) {
			COLOR_GRADATIONS = new IntMap<LGradation>(10);
		}
		int hashCode = 1;
		hashCode = LSystem.unite(hashCode, s.getRGB());
		hashCode = LSystem.unite(hashCode, e.getRGB());
		hashCode = LSystem.unite(hashCode, w);
		hashCode = LSystem.unite(hashCode, h);
		hashCode = LSystem.unite(hashCode, alpha);
		LGradation o = COLOR_GRADATIONS.get(hashCode);
		if (o == null) {
			COLOR_GRADATIONS.put(hashCode, o = new LGradation(s, e, w, h, alpha));
		}
		return o;
	}

	private LGradation() {

	}

	private LGradation(LColor s, LColor e, int w, int h, int alpha) {
		this.startColor = s;
		this.endColor = e;
		this.width = w;
		this.height = h;
		this.alpha = alpha;
	}

	public void drawWidth(GLEx g, int x, int y) {
		try {
			if (drawTexWidth == null && !drawTexWidth.isClosed()) {
				Canvas gl = LSystem.base().graphics().createCanvas(width, height);
				for (int i = 0; i < width; i++) {
					gl.setColor((startColor.getRed() * (width - i)) / width + (endColor.getRed() * i) / width,
							(startColor.getGreen() * (width - i)) / width + (endColor.getGreen() * i) / width,
							(startColor.getBlue() * (width - i)) / width + (endColor.getBlue() * i) / width, alpha);
					gl.drawLine(i, 0, i, height);
				}
				drawTexWidth = gl.toTexture();
				if (gl.image != null) {
					gl.image.close();
				}
			}
			g.draw(drawTexWidth, x, y);
		} catch (Throwable ex) {
			for (int i = 0; i < width; i++) {
				g.setColor((startColor.getRed() * (width - i)) / width + (endColor.getRed() * i) / width,
						(startColor.getGreen() * (width - i)) / width + (endColor.getGreen() * i) / width,
						(startColor.getBlue() * (width - i)) / width + (endColor.getBlue() * i) / width, alpha);
				g.drawLine(i + x, y, i + x, y + height);
			}
		}
	}

	public void drawHeight(GLEx g, int x, int y) {
		try {
			if (drawTexHeight == null && !drawTexHeight.isClosed()) {
				Canvas gl = LSystem.base().graphics().createCanvas(width, height);
				for (int i = 0; i < height; i++) {
					gl.setColor((startColor.getRed() * (height - i)) / height + (endColor.getRed() * i) / height,
							(startColor.getGreen() * (height - i)) / height + (endColor.getGreen() * i) / height,
							(startColor.getBlue() * (height - i)) / height + (endColor.getBlue() * i) / height, alpha);
					gl.drawLine(0, i, width, i);
				}
				drawTexHeight = gl.toTexture();
				if (gl.image != null) {
					gl.image.close();
				}
			}
			g.draw(drawTexHeight, x, y);
		} catch (Throwable ex) {
			for (int i = 0; i < height; i++) {
				g.setColor((startColor.getRed() * (height - i)) / height + (endColor.getRed() * i) / height,
						(startColor.getGreen() * (height - i)) / height + (endColor.getGreen() * i) / height,
						(startColor.getBlue() * (height - i)) / height + (endColor.getBlue() * i) / height, alpha);
				g.drawLine(x, i + y, x + width, i + y);
			}
		}
	}

	public void drawWidth(Canvas g, int x, int y) {
		try {
			if (drawImgWidth == null && !drawImgWidth.isClosed()) {
				Canvas gl = LSystem.base().graphics().createCanvas(width, height);
				drawImgWidth = gl.image;
				for (int i = 0; i < width; i++) {
					gl.setColor((startColor.getRed() * (width - i)) / width + (endColor.getRed() * i) / width,
							(startColor.getGreen() * (width - i)) / width + (endColor.getGreen() * i) / width,
							(startColor.getBlue() * (width - i)) / width + (endColor.getBlue() * i) / width, alpha);
					gl.drawLine(i, 0, i, height);
				}
				gl.close();
				gl = null;
			}
			g.draw(drawImgWidth, x, y);
		} catch (Throwable e) {
			for (int i = 0; i < width; i++) {
				g.setColor((startColor.getRed() * (width - i)) / width + (endColor.getRed() * i) / width,
						(startColor.getGreen() * (width - i)) / width + (endColor.getGreen() * i) / width,
						(startColor.getBlue() * (width - i)) / width + (endColor.getBlue() * i) / width, alpha);
				g.drawLine(i + x, y, i + x, y + height);
			}
		}
	}

	public void drawHeight(Canvas g, int x, int y) {
		try {
			if (drawImgHeight == null && !drawImgHeight.isClosed()) {
				Canvas gl = LSystem.base().graphics().createCanvas(width, height);
				drawImgHeight = gl.image;
				for (int i = 0; i < height; i++) {
					gl.setColor((startColor.getRed() * (height - i)) / height + (endColor.getRed() * i) / height,
							(startColor.getGreen() * (height - i)) / height + (endColor.getGreen() * i) / height,
							(startColor.getBlue() * (height - i)) / height + (endColor.getBlue() * i) / height, alpha);
					gl.drawLine(0, i, width, i);
				}
				gl.close();
				gl = null;
			}
			g.draw(drawImgHeight, x, y);
		} catch (Throwable e) {
			for (int i = 0; i < height; i++) {
				g.setColor((startColor.getRed() * (height - i)) / height + (endColor.getRed() * i) / height,
						(startColor.getGreen() * (height - i)) / height + (endColor.getGreen() * i) / height,
						(startColor.getBlue() * (height - i)) / height + (endColor.getBlue() * i) / height, alpha);
				g.drawLine(x, i + y, x + width, i + y);
			}
		}
	}

	public void drawWidth(Pixmap g, int x, int y) {
		try {
			if (drawImgWidth == null && !drawImgWidth.isClosed()) {
				for (int i = 0; i < width; i++) {
					g.setColor((startColor.getRed() * (width - i)) / width + (endColor.getRed() * i) / width,
							(startColor.getGreen() * (width - i)) / width + (endColor.getGreen() * i) / width,
							(startColor.getBlue() * (width - i)) / width + (endColor.getBlue() * i) / width, alpha);
					g.drawLine(i, 0, i, height);
				}
				drawImgWidth = g.getImage();
			}
		} catch (Throwable e) {
			for (int i = 0; i < width; i++) {
				g.setColor((startColor.getRed() * (width - i)) / width + (endColor.getRed() * i) / width,
						(startColor.getGreen() * (width - i)) / width + (endColor.getGreen() * i) / width,
						(startColor.getBlue() * (width - i)) / width + (endColor.getBlue() * i) / width, alpha);
				g.drawLine(i + x, y, i + x, y + height);
			}
		}
	}

	public void drawHeight(Pixmap g, int x, int y) {
		try {
			if (drawImgHeight == null && !drawImgHeight.isClosed()) {
				for (int i = 0; i < height; i++) {
					g.setColor((startColor.getRed() * (height - i)) / height + (endColor.getRed() * i) / height,
							(startColor.getGreen() * (height - i)) / height + (endColor.getGreen() * i) / height,
							(startColor.getBlue() * (height - i)) / height + (endColor.getBlue() * i) / height, alpha);
					g.drawLine(0, i, width, i);
				}
				drawImgHeight = g.getImage();
			}
		} catch (Throwable e) {
			for (int i = 0; i < height; i++) {
				g.setColor((startColor.getRed() * (height - i)) / height + (endColor.getRed() * i) / height,
						(startColor.getGreen() * (height - i)) / height + (endColor.getGreen() * i) / height,
						(startColor.getBlue() * (height - i)) / height + (endColor.getBlue() * i) / height, alpha);
				g.drawLine(x, i + y, x + width, i + y);
			}
		}
	}

	public static void dispose() {
		if (COLOR_GRADATIONS == null) {
			return;
		}
		for (LGradation g : COLOR_GRADATIONS.values()) {
			if (g != null) {
				g.close();
			}
		}
		COLOR_GRADATIONS.clear();
	}

	@Override
	public void close() {
		if (drawTexWidth != null) {
			drawTexWidth.close();
		}
		if (drawTexHeight != null) {
			drawTexHeight.close();
		}
		if (drawImgWidth != null) {
			drawImgWidth.close();
		}
		if (drawImgWidth != null) {
			drawImgWidth.close();
		}
	}

}
